src/org/openstreetmap/josm/data/validation/tests/TagChecker.java

```diff
@@ -654,15 +663,21 @@ public void check(OsmPrimitive p) {
         checkMultipolygonTags(p);
         }
 
+        final Collection<TaggingPreset> matchingPresets;
+        TagMap tags = p.getKeys();
+        if (checkPresetsTypes || checkRegions) {
+            matchingPresets = presetIndex.entrySet().stream()
+                    .filter(e -> TaggingPresetItem.matches(e.getValue(), tags))
+                    .map(Entry::getKey)
+                    .collect(Collectors.toCollection(LinkedHashSet::new));
```

Comment:
I know you just copied and pasted, but have you considered _why_ `LinkedHashSet` is used instead of the more standard `ArrayList`?

Reply:
One thing I know about `HashSet` is that it doesn't allow duplicate entries, which is not the case with `ArrayList`.

I had to search for `LinkedHashSet`. It maintains the order of the inserted elements. Is this the reason why `LinkedHashSet` has been used?

Reply:
You are correct in that
- List maintains insert order, but allows duplicates
- Set doesn't allow duplicates, but doesn't guarantee that insertion order will be maintained

LinkedHashSet does maintain insertion oder and deduplication, but that comes at the cost of internally being quite complex, so this should only be used when both qualities are actually required.

Reply:
I suspect that `LinkedHashSet` was used mostly for deduplication. It was initially added to the code in [r15788](https://josm.openstreetmap.de/changeset/15788/josm/).

I don't believe any of the code written depends upon the _order_ of that list, so either a standard `List` class (with `.distinct()` in the stream) will work _or_ a standard `Set` class will work.

In this case, it isn't terribly important since we aren't passing the set around, but it is in code that can be run millions of times in a loop (if someone decides to validate a country, for example).

IIRC, you are using IntelliJ. If you ever encounter code that you do not understand the purpose of, you can right-click in the "gutter" and enable annotations with `Annotate`. This will then let you click on a line in the gutter to see the commit information. You can also right-click on the line in the gutter to annotate a _previous_ revision, which can be really handy when someone decides to bulk edit stuff.